lib/logstash/codecs/identity_map_codec.rb

```diff
@@ -0,0 +1,176 @@
+# encoding: utf-8
+require "logstash/namespace"
+require "concurrent"
+
+# This class is a Codec duck type
+# Using Composition, it maps from a stream identity to
+# a cloned codec instance via the same API as a Codec
+
+module LogStash module Codecs class IdentityMapCodec
+  # subclass of Exception, LS has more than limit (100) active streams
+  class IdentityMapUpperLimitException < Exception; end
+
+  module EightyPercentWarning
+    extend self
+    def visit(imc)
+      current_size, limit = imc.current_size_and_limit
+      return if current_size < (limit * 0.8)
+      imc.logger.warn("IdentityMapCodec has reached 80% capacity",
+        :current_size => current_size, :upper_limit => limit)
+    end
+  end
+
+  module UpperLimitReached
+    extend self
+    def visit(imc)
+      current_size, limit = imc.current_size_and_limit
+      return if current_size < limit
+      current_size, limit = imc.map_cleanup
+      return if current_size < limit
+      imc.logger.error("IdentityMapCodec has reached 100% capacity",
+          :current_size => current_size, :upper_limit => limit)
+      raise IdentityMapUpperLimitException.new
+    end
+  end
+
+  class MapCleaner
+    def initialize(imc, interval)
+      @running = true
+      @imc, @interval = imc, interval
+    end
+
+    def run
+      @thread = Thread.new(@imc) do |imc|
+        loop do
+          sleep @interval
+          break if !@running
+          imc.map_cleanup
+        end
+      end
+      self
+    end
+
+    def stop
+      @running = false
+      @thread.wakeup
+    end
+  end
+
+  MAX_IDENTITIES = 100
+  EVICT_TIMEOUT = 60 * 60 * 4 # 4 hours
+  CLEANER_INTERVAL = 60 * 5 # 5 minutes
+
+  attr_reader :identity_map, :usage_map
+  attr_accessor :base_codec, :logger, :cleaner
+
+  def initialize(codec, logger)
+    @base_codec = codec
+    @base_values = [codec]
+    @identity_map = Hash.new &method(:codec_builder)
```

Comment:
I do this because passing a block is usually done outside the parentheses.  It is a syntactic anomaly in Ruby.
```ruby
method_call(42) {|arg| some_code }
method_call(42) do |arg|
  some_code
end
method_call(42, &method(:sym))
method_call 42, &method(:sym)
# whereas these fail
method_call(42) &method(:sym)
method_call 42, {|arg| some_code}
```
But we never write `array.<<(member)` or `60.*(60).*(4)` or `hash.[]=('key', 42)`